### DISCUSSION TITLE
在Jupyter上驗證功能的方法

### DISCUSSION THREAD

**Discussion by Fossum2523** [bol-edu/HLS-SOC-Discussions#193] [Q&A]
要如何在Jupyter上測試MM、QS、FIR的輸出資料是否正確呢?

因為在收取online FPGA回傳到Jupyter的資料時會發現很容易掉包，我猜測是因為計算的速度很快，所以Jupyter端去fetch值得時候資料早就換了。

我們在先前Lab 6當中測試功能的時候在每筆資料的輸出後加上延遲，使得Jupyter端可以fetch到資料，如下圖。
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/131112486/00562d98-92d0-4f6a-9340-376893269c03)

但是老師有提到過這算是偷吃步，故想來請教各位是如何驗證的呢?

目前想到的方法是先由Jupyter端打一個要收資料的訊號過去，FPGA收到後再將資料回傳回來，但這方法我一直嘗試不出來。

**Comment by billy900812** [comment]
我也有想嘗試你的這個方法，但我去看了GPIO的相關資料好像是做不到，如下圖給你參考
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/155391377/caf2628d-7d7c-40e5-9b1a-78c852b6bc00)

**Comment by boledulab** [comment]
When Caravel SOC is integrated in FPGA, its bidirectional mprj pins are separated into input, output, and output enable signals (figure below) which can be read/write by PS. Refer to the caravel_ps HLS code below. Unlike the bidirectional pin in real chip, it has separate and independent input and output signals. It can be used for Firmware and PS communication. 
Firmware can read write mprj registers, and PS can read / write the axilite registers.  You may use this to do lock-step operations between firmware and PS. i.e. 
1. firmware waits for PS to send a message, then starts computation
2. After the computation is done, send the result back to PS
3. PS read the result, then send a message to the firmware. 
4. Repeat 1-3.

![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/98332019/ddc257bd-f764-4dc6-b321-4e7e902bc55e)

Refer to the following for Caravel_PS HLS code.  The mprj_in, mprj_out
https://github.com/bol-edu/caravel-soc_fpga-lab/blob/main/labi/vitis_hls_project/hls_caravel_ps/src/caravel_ps.cpp

**Reply by CatKingWoof** [reply]
老師您好，請問final project有需要驗證workload計算的每個結果是否正確嗎?

**Reply by boledulab** [reply]
需要！
